helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -241,13 +244,87 @@ private void refreshClusterConfig(final HelixDataAccessor accessor,
     if (_propertyDataChangedMap.get(HelixConstants.ChangeType.CLUSTER_CONFIG).getAndSet(false)) {
       _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
       refreshedType.add(HelixConstants.ChangeType.CLUSTER_CONFIG);
+      // TODO: This is a temp function to clean up incompatible batched disabled instances format.
+      // Remove in later version.
+      if (checkBatchedDisabledInstanceFormat(_clusterConfig) && updateBatchDisableFormat(
+          accessor)) {
+        // read from zkz one more time
+        LogUtil.logInfo(logger, getClusterEventId(), String
+            .format("Clean ClusterConfig change for cluster %s, pipeline %s", _clusterName,
+                getPipelineName()));
+        _clusterConfig = accessor.getProperty(accessor.keyBuilder().clusterConfig());
```

Comment:
Let's not read this again. Helix computation is based on a snapshot of all data for ZK clusters. If you read cluster config in the middle of the pipeline, then cluster config is a newer version of snapshot instead of original one.

My suggestion is just reformat whatever we need to update in memory. But do not read cluster config back. Let next round pipe line refresh to refresh cluster config since there should be such type of config change notification.

Reply:
Second to Junkai, the immediate _clusterConfig refresh may bring in other changes in the middle of the pipeline. It will become tricky to tell when a clusterConfig cache is refreshed. 
The format change isn't a substantial one, I don't think it will cause any new issue in the pipeline (assuming it was in a good state). Could you please confirm?

Reply:
TFTR.
Waged will read disabled time in BestPossibleComputeState. If we keep the uncleaned format, it will cause issue for the pipeline run.
I think the question is, is it ok for us to have several failed pipeline until the change in cluster config is picked up?

Reply:
I think what Junkai is suggesting is instead of writing corrected data to ZK in `updateBatchDisableFormat()`, and then reading it back in `_clusterConfig` here, directly update `_clusterConfig` in `updateBatchDisableFormat()` and let the data written to ZK be picked up in `_clusterConfig` the next time pipeline is triggered. This way, we won't have the risk of pipeline failure due to incorrect disabled time format, or the risk of inconsistent `_clusterConfig` between pipeline stages.

Reply:
Thanks for the review.
Please correct me if I am wrong, if update failed then we should not update ` _clusterConfig ` ?

Reply:
That's a good question, if we know the format is wrong and we fail to update ZK, do we want to keep the wrong format in cache as well? My personal opinion is that the cache should be fixed regardless, but I might be missing something. Regardless, I think you can check the result of the update and based on that update `_clusterConfig`. Would that cause duplicate work?

Reply:
In my understanding, the baseline is that whatever in cache should be the same as ZNode. If we failed to update ZNode and the format causing pipeline fail, then the pipeline fail and we try again.  
I am open to discussion.

Reply:
In general, I agree that cache and ZK must have the same view. However, in this case what's the risk of updating cache ahead of ZK? It's not new information, but just fixing formatting. The same fix to ZK will be attempted the next time the pipeline is running. But even if we want to avoid version mismatch, the success of ZK update can be checked and then instead of reading from ZK again, the same changes can be applied to `_clusterConfig`.

Reply:
Thanks for suggestion. Updated in follow up commit.